core/src/main/scala/org/apache/spark/scheduler/AsyncEventQueue.scala

```diff
@@ -142,10 +142,17 @@ private class AsyncEventQueue(
       eventCount.incrementAndGet()
       eventQueue.put(POISON_PILL)
     }
-    // this thread might be trying to stop itself as part of error handling -- we can't join
-    // in that case.
-    if (Thread.currentThread() != dispatchThread) {
-      dispatchThread.join()
+    // 1. This thread might be trying to stop itself as part of error handling -- we can't join
+    //    in that case.
+    // 2. If users don't want to wait for the dispatch to end until all events are drained,
+    //    they can control the waiting time by themselves
+    //    or omit the thread join by set the waiting time to a negative value.
+    val waitingTimeMs =
+      conf.get(LISTENER_BUS_EVENT_QUEUE_EVENT_DISPATCH_EXIT_WAITING_TIME_ON_STOP)
+    if (waitingTimeMs >= 0 && Thread.currentThread() != dispatchThread) {
+      // By default, the `waitingTimeMs` is set to 0,
+      // which means it will wait until all events are drained.
+      dispatchThread.join(waitingTimeMs)
     }
```

Comment:
```suggestion
    if (Thread.currentThread() != dispatchThread) {
      // By default, the `waitingTimeMs` is set to 0,
      // which means it will wait until all events are drained.
      dispatchThread.join(waitingTimeMs)
    }
```